polynote-frontend/polynote/ui/component/messaging/receiver.ts

```diff
@@ -0,0 +1,331 @@
+/**
+ * The MessageReceiver is used to translate external events into state changes.
+ * So far, the only external events are socket messages.
+ */
+import {SocketSession} from "../../../comms";
+import {CellState, NotebookState, NotebookStateHandler} from "../state/notebook_state";
+import {ServerState, ServerStateHandler} from "../state/server_state";
+import * as messages from "../../../data/messages";
+import {TaskInfo} from "../../../data/messages";
+import {CellComment, CellMetadata, NotebookCell, NotebookConfig} from "../../../data/data";
+import match from "../../../util/match";
+import {Message} from "../../../data/messages";
+import {ContentEdit} from "../../../data/content_edit";
+import {
+    ClearResults, ClientResult,
+    CompileErrors,
+    ExecutionInfo,
+    Output,
+    PosRange,
+    ResultValue,
+    RuntimeError
+} from "../../../data/result";
+import {StateHandler} from "../state/state_handler";
+import {clientInterpreters} from "../../../interpreter/client_interpreter";
+import {EditBuffer} from "../../../data/edit_buffer";
+
+class MessageReceiver<S> {
```

Comment:
These MessageReceivers contain the logic that reads messages from the socket and translates them into updates to the state.

Reply:
they're 1:1 with sockets - one for the server socket, one for each notebook socket